src/pages/DeviceSensor.tsx

```diff
@@ -0,0 +1,217 @@
+import { Box, Typography, Breadcrumbs } from "@mui/material";
+import RowContainerBetween from "../components/shared/RowContainerBetween";
+import { useLocation, useNavigate, useOutletContext, useParams } from "react-router-dom";
+import Chart from 'react-apexcharts';
+import { useCallback, useLayoutEffect, useState } from "react";
+import type { Actuator, Device, Sensor } from "waziup";
+import { Link } from "react-router-dom";
+import PrimaryIconButton from "../components/shared/PrimaryIconButton";
+import SensorTable from "../components/ui/DeviceTable";
+import { cleanString } from "../utils";
+function Device() {
+    function handleClick(event: React.MouseEvent<Element, MouseEvent>) {
+        event.preventDefault();
+    }
+    const { pathname } = useLocation();
+    const [device, setDevice] = useState<Device | null>(null);
+    const [sensOrActuator, setSensOrActuator] = useState<Sensor | Actuator | null>(null)
+    const [values, setValues] = useState<{ value: number | string, modified: string }[]>([]);
+    const [valsLimit, setValsLimit] = useState<number>(700);
+    const [graphValues, setGraphValues] = useState<{ y: number, x: string }[]>([]);
+    const [matches] = useOutletContext<[matches: boolean]>();
+
+    const navigate = useNavigate();
+    const { id, sensorId } = useParams();
+    const getGraphValues = useCallback(function (deviceId: string, sensorId: string) {
```

Comment:
This function is strange. It is named "getGraphValues", but it also sets the values for the table.
I recommend:
1. Fetch the sensors values and put them in a local variable (not formatting, just fetching and storing)
2. Provide 2 functions for tranforming the data: one for the graph, one for the table
3. Use those functions "on the fly" when building the graph and table. No need to store the intermediate results.